dkg/src/dkg_manager/mod.rs

```diff
@@ -40,29 +95,199 @@ impl<DKG: DKGTrait, P: DKGPrivateParamsProvider<DKG>> DKGManager<DKG, P> {
             vtxn_pool_write_cli,
             agg_trx_tx: None,
             agg_trx_producer,
+            stopped: false,
+            state: InnerState::NotStarted,
         }
     }
 
     pub async fn run(
-        self,
-        _in_progress_session: Option<DKGSessionState>,
-        _dkg_start_event_rx: aptos_channel::Receiver<(), DKGStartEvent>,
-        _rpc_msg_rx: aptos_channel::Receiver<(), (AccountAddress, IncomingRpcRequest)>,
-        _dkg_txn_pulled_rx: vtxn_pool::PullNotificationReceiver,
+        mut self,
+        in_progress_session: Option<DKGSessionState>,
+        mut dkg_start_event_rx: aptos_channel::Receiver<(), DKGStartEvent>,
+        mut rpc_msg_rx: aptos_channel::Receiver<(), (AccountAddress, IncomingRpcRequest)>,
+        mut dkg_txn_pulled_rx: vtxn_pool::PullNotificationReceiver,
         close_rx: oneshot::Receiver<oneshot::Sender<()>>,
     ) {
+        let (agg_trx_tx, mut agg_trx_rx) = aptos_channel::new(QueueStyle::KLAST, 1, None);
+        self.agg_trx_tx = Some(agg_trx_tx);
+
         let mut close_rx = close_rx.into_stream();
-        loop {
-            tokio::select! {
-                //TODO: handle other events
+        while !self.stopped {
+            let handling_result = tokio::select! {
+                dkg_start_event = dkg_start_event_rx.select_next_some() => {
+                    self.process_dkg_start_event(dkg_start_event).await
+                },
+                (_sender, msg) = rpc_msg_rx.select_next_some() => {
+                    self.process_peer_rpc_msg(msg).await
+                },
+                agg_node = agg_trx_rx.select_next_some() => {
+                    self.process_aggregated_transcript(agg_node).await
+                },
+                dkg_txn = dkg_txn_pulled_rx.select_next_some() => {
+                    self.process_dkg_txn_pulled_notification(dkg_txn).await
+                },
                 close_req = close_rx.select_next_some() => {
-                    self.vtxn_pool_write_cli.put(None);
-                    if let Ok(ack_sender) = close_req {
-                        ack_sender.send(()).unwrap();
-                    }
-                    break;
+                    self.process_close_cmd(close_req.unwrap())
```

Comment:
nit: I think we should just tolerate the close_req being Err (the sender was dropped) for sanity

Reply:
fixed